### PR TITLE
Only read input once

### PR DESCRIPTION
**Pull request by rikonor**
I was having trouble using stdin/stdout for input/output to jpeg-recompress.

Digging a little bit deeper and with the help of @pope I realized that jpeg-recompress actually reads the input multiple times, so by the second read the input is already drained. This is why it works fine with files, but not with pipes/stdin.

The fix adds new `decodeFiletype` and `decodeFile` versions that take a buffer instead of reading the file themselves.

### CODE REVIEW

File: src/util.c [hunk 1]
```diff
@@ -276,6 +276,17 @@ enum filetype detectFiletype(const char *filename) {
     return ret;
```
Comment (pope) [inline src/util.c]: Consider calling detectFiletypeFromBuffer here:

```c
enum filetype detectFiletype(const char *filename) {
     unsigned char *buf = NULL;
     long bufSize = 0;
     bufSize = readFile((char *)filename, (void **)&buf);
     enum filetype  ret = detectFiletypeFromBuffer(buf, bufSize);
     free(buf);
     return ret;
}
```

File: src/util.c [hunk 2]
```diff
@@ -297,6 +308,22 @@ unsigned long decodeFile(const char *filename, unsigned char **image, enum filet
...
 }
 
+unsigned long decodeFileFromBuffer(unsigned char *buf, long bufSize, unsigned char **image, enum filetype type, int *width, int *height, int pixelFormat) {
```
Comment (pope) [inline src/util.c]: ditto that `decodeFile` should call `decodeFileFromBuffer`

File: src/util.c [hunk 3]
```diff
@@ -276,6 +276,17 @@ enum filetype detectFiletype(const char *filename) {
     return ret;
 }
 
+enum filetype detectFiletypeFromBuffer(unsigned char *buf, long bufSize) {
```
Comment (pope) [inline src/util.c]: You can simplify:

```c
if (checkJpegMagic(buf, bufSize))
    return FILETYPE_JPEG;

if (checkPpmMagic(buf, bufSize))
    return FILETYPE_PPM;

return FILETYPE_UNKNOWN;
```

File: src/util.c [hunk 4]
```diff
@@ -297,6 +308,22 @@ unsigned long decodeFile(const char *filename, unsigned char **image, enum filet
     return ret;
 }
 
+unsigned long decodeFileFromBuffer(unsigned char *buf, long bufSize, unsigned char **image, enum filetype type, int *width, int *height, int pixelFormat) {
+    long ret = 0;
+
+    if (!bufSize)
+        return 0;
+
+    if (type == FILETYPE_PPM)
```
Comment (pope) [inline src/util.c]: You can simplify:

```c
switch (type) {
  case FILETYPE_PPM:
    return decodePpm(buf, bufSize, image, width, height);
  case FILETYPE_JPEG:
    return decodeJpeg(buf, bufSize, image, width, height, pixelFormat);
  default:
    return 0;
}
```

File: src/util.c (hunk 1 above)
Comment (rikonor) [reply src/util.c]: Done

File: src/util.c (hunk 2 above)
Comment (rikonor) [reply src/util.c]: Done

File: src/util.c (hunk 3 above)
Comment (rikonor) [reply src/util.c]: Done

File: src/util.c (hunk 4 above)
Comment (rikonor) [reply src/util.c]: Done

File: jpeg-recompress.c [hunk 5]
```diff
@@ -322,16 +322,21 @@ int main (int argc, char **argv) {
         setTargetFromPreset();
     }
 
+    char *inputPath = (char *) cmd.argv[0];
+    char *outputPath = cmd.argv[1];
+
+    // Read the input into a buffer
```
Comment (lfos) [inline jpeg-recompress.c]: Small nit: Please use C-style comments (`/* */`) here.

File: .gitignore [hunk 6]
```diff
@@ -10,3 +10,4 @@ comparison
 comparison.png
 test/test.jpg
 test/test
+.DS_Store
```
Comment (lfos) [inline .gitignore]: This seems to be unrelated? Please drop.

File: src/util.c [hunk 7]
```diff
@@ -263,65 +263,40 @@ unsigned long decodePpm(unsigned char *buf, unsigned long bufSize, unsigned char
...
-
-    if (type == FILETYPE_PPM)
-        ret = decodePpm(buf, bufSize, image, width, height);
-    else if (type == FILETYPE_JPEG)
-        ret = decodeJpeg(buf, bufSize, image, width, height, pixelFormat);
-    else
-        ret = 0;
-
-    return ret;
+    switch(type) {
```
Comment (lfos) [inline src/util.c]: Style nit: Missing space after `switch`.

File: src/util.h [hunk 8]
```diff
@@ -58,9 +58,11 @@ unsigned long encodeJpeg(unsigned char **jpeg, unsigned char *buf, int width, in
 
 /* Automatically detect the file type of a given file. */
 enum filetype detectFiletype(const char *filename);
+enum filetype detectFiletypeFromBuffer(unsigned char *buf, long bufSize);
```
Comment (lfos) [inline src/util.h]: Should this be part of the first patch, where the new function is introduced? Same for `decodeFileFromBuffer` below.

**Comment by rikonor** [comment]
Hey @lfos, thanks for the review.

I'm sorry but I won't have time to touch this in the next few weeks. If you're open to it, I'm fine with you taking over the PR and making changes as you see fit.

**Comment by lfos** [comment]
Merged to `pu` (with some minor changes). Thanks!
